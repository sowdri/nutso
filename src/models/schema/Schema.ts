import { ArraySchema } from "./ArraySchema";
import { BooleanSchema } from "./BooleanSchema";
import { DateSchema } from "./DateSchema";
import { NumberSchema } from "./NumberSchema";
import { ObjectSchema } from "./ObjectSchema";
import { StringSchema } from "./StringSchema";
import { TupleSchema } from "./TupleSchema";

/**
 * # Terminology (Only required to refer the code)
 * T => Type of the value currently being validated
 * K => key => Type of Key
 * R => root => The root of the object that is being validated
 * P => parent => The parent of the object being validated. If for [K in keyof T],
 */

export type Schema<T, R = T, P = unknown> = T extends string
  ? StringSchema<R, P>
  : T extends number
  ? NumberSchema<R, P>
  : T extends Date
  ? DateSchema<R, P>
  : T extends boolean
  ? BooleanSchema<R, P>
  : T extends Array<infer E>
  ? ArraySchema<E, T, R, P>
  : T extends [infer U, ...unknown[]]
  ? TupleSchema<U, R, P>
  : T extends Object
  ? ObjectSchema<T, R, P>
  : never;
