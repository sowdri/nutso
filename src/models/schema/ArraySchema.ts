import { OptionalFlag } from "../OptionalFlag";
import { Schema } from "./Schema";

/**
 * E => Element
 * T => Type (this is the array)
 * P => Parent of the array
 *
 * Look at this type, we need all 3 types
 */
export type ArraySchema<E, T extends E[], R, P = unknown> = {
  type: "array";
  optional?: OptionalFlag<R, P>;
  minItems?: number;
  maxItems?: number;
  items: Schema<E, R>; // TODO check how this works in schema definition
};
