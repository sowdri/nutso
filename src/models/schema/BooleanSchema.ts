import { OptionalFlag } from "../OptionalFlag";
import { ValidationFn } from "../ValidationFn";

export type BooleanSchema<R = boolean, P = unknown> = {
  type: "boolean";
  optional?: OptionalFlag<R, P>;
  validationFn?: ValidationFn<boolean, R, P>;
};
