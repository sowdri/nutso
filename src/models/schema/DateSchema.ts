import { OptionalFlag } from "../OptionalFlag";
import { ValidationFn } from "../ValidationFn";

export type DateSchema<R = Date, P = unknown> = {
  type: "date";
  optional?: OptionalFlag<R, P>;
  validationFn?: ValidationFn<Date, R, P>;
};
