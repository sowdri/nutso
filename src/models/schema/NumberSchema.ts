import { OptionalFlag } from "../OptionalFlag";
import { ValidationFn } from "../ValidationFn";

export type NumberSchema<R = number, P = unknown> = {
  type: "number";
  min?: number;
  max?: number;
  pattern?: RegExp;
  optional?: OptionalFlag<R, P>;
  validationFn?: ValidationFn<number, R, P>;
};
