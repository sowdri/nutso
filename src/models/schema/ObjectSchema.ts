import { Schema } from "./Schema";
import { OptionalFlag } from "../OptionalFlag";

export type ObjectSchema<T, R = T, P = unknown> = {
  type: "object";
  optional?: OptionalFlag<R, P>;
  properties: {
    [K in keyof T]: Schema<T[K], R, T>;
  };
};
