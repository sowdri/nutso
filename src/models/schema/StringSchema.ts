import { OptionalFlag } from "../OptionalFlag";
import { ValidationFn } from "../ValidationFn";
import { Schema } from "./Schema";

/*
type ValidatorFnResult1 = {
  errorMessage: string;
};

type ValidationFn1<T, R, P> = (args: { value: T; parent: P; root: R }) => ValidatorFnResult1 | void;

*/
export type StringSchema<R = string, P = unknown> = {
  type: "string";
  optional?: OptionalFlag<R, P>;
  minLength?: number;
  maxLength?: number;
  pattern?: RegExp;
  validationFn?: ValidationFn<string, R, P>;
};

const nakedSchema: Schema<string> = {
  type: "string",
  validationFn: (args) => {},
};

type Customer = {
  name: string;
};

const schema: Schema<Customer> = {
  type: "object",
  properties: {
    name: {
      type: "string",
      validationFn: (args) => {},
    },
  },
};

type Bar = {
  bar: string;
};

type Foo = {
  bar: Bar;
};

const fooSchema: Schema<Foo> = {
  type: "object",
  properties: {
    bar: {
      type: "object",
      properties: {
        bar: {
          type: "string",
          validationFn: (args) => {},
        },
      },
    },
  },
};
