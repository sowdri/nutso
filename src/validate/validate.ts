import { Result } from "../models/result/Result";
import { BooleanSchema } from "../models/schema/BooleanSchema";
import { DateSchema } from "../models/schema/DateSchema";
import { NumberSchema } from "../models/schema/NumberSchema";
import { Schema } from "../models/schema/Schema";
import { StringSchema } from "../models/schema/StringSchema";
import { validateArray } from "./validateArray";
import { validateBoolean } from "./validateBoolean";
import { validateDate } from "./validateDate";
import { validateNumber } from "./validateNumber";
import { validateObject } from "./validateObject";
import { validateString } from "./validateString";

/**
 * This is an internal API that takes the root as an argument
 * and called recursively from `validateObject` and `validateArray` methods
 *
 * @param value
 * @param parent
 * @param schema
 */
export const _validate = <T, P>(args: { value: T | null; parent: P; schema: Schema<T, P> }): Result<T> => {
  //

  if (!args.schema) {
    throw new Error("Schema should not be null");
  }

  /**
   * validate object against the schema
   */
  switch (args.schema.type) {
    case "number":
      return validateNumber({ ...args, schema: args.schema as NumberSchema<P> }) as Result<T>;
    case "string":
      return validateString({ ...args, schema: args.schema as StringSchema<P> }) as Result<T>;
    case "boolean":
      return validateBoolean({ ...args, schema: args.schema as BooleanSchema<P> }) as Result<T>;
    case "array":
      return validateArray({ value: args.value as any, parent: args.parent, schema: args.schema as any }) as Result<T>;
    case "object":
      return validateObject({ ...args, schema: args.schema }) as Result<T>;
    case "date":
      return validateDate({ ...args, schema: args.schema as DateSchema<P> }) as Result<T>;
  }
  throw new Error(`Unhandled data type`);
};

// public api
export const validate = <T>(value: T, schema: Schema<T>): Result<T> => {
  return _validate({ value, parent: null as any, schema });
};
