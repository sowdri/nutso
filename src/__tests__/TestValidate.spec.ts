import { Schema } from "../models/schema/Schema";
import { validate } from "../validate/validate";

test(`Test customer schema`, () => {
  interface Customer {
    name: string;
    address: Address;
  }
  interface Address {
    line1: string;
  }
  const customerSchema: Schema<Customer> = {
    type: "object",
    optional: false,
    properties: {
      name: {
        type: "string",
        minLength: 3,
        maxLength: 24,
      },
      address: {
        type: "object",
        properties: {
          line1: {
            type: "string",
          },
        },
      },
    },
  };
  const customer: Customer = {
    name: "nish",
    address: {
      line1: "18/19 Lightsview",
    },
  };
  expect(validate(customer, customerSchema)).toMatchSnapshot();
});

test(`Test index type`, () => {
  type Car = {
    vin: string;
  };

  interface Customer {
    name: string;
    address: Address;
    cars: {
      [key: string]: Car;
    };
  }
  interface Address {
    line1: string;
  }
  const customerSchema: Schema<Customer> = {
    type: "object",
    optional: false,
    properties: {
      name: {
        type: "string",
        minLength: 3,
        maxLength: 24,
      },
      address: {
        type: "object",
        properties: {
          line1: {
            type: "string",
          },
        },
      },
      cars: {
        type: "object",
        properties: {
          rugby: {
            type: "object",
            properties: {
              vin: {
                type: "string",
              },
            },
          },
        },
      },
    },
  };
  const customer: Customer = {
    name: "nish",
    address: {
      line1: "18/19 Lightsview",
    },
    cars: {
      rugby: {
        vin: "xsb797",
      },
    },
  };
  const result = validate(customer, customerSchema);
  expect(result).toMatchSnapshot();
});

test(`Test index type - required field - missing`, () => {
  type Car = {
    vin: string;
  };

  interface Customer {
    name: string;
    address: Address;
    cars: {
      [key: string]: Car;
    };
  }
  interface Address {
    line1: string;
  }
  const customerSchema: Schema<Customer> = {
    type: "object",
    optional: false,
    properties: {
      name: {
        type: "string",
        minLength: 3,
        maxLength: 24,
      },
      address: {
        type: "object",
        properties: {
          line1: {
            type: "string",
          },
        },
      },
      cars: {
        type: "object",
        properties: {
          // not, it's rugby in the obj
          ruby: {
            type: "object",
            properties: {
              vin: {
                type: "string",
              },
            },
          },
        },
      },
    },
  };
  const customer: Customer = {
    name: "nish",
    address: {
      line1: "18/19 Lightsview",
    },
    cars: {
      rugby: {
        vin: "xsb797",
      },
    },
  };
  const result = validate(customer, customerSchema);
  expect(result).toMatchSnapshot();
});

test(`Test index type - regex`, () => {
  type Car = {
    vin: string;
  };

  interface Customer {
    name: string;
    address: Address;
    cars: {
      [key: string]: Car;
    };
  }
  interface Address {
    line1: string;
  }
  const customerSchema: Schema<Customer> = {
    type: "object",
    optional: false,
    properties: {
      name: {
        type: "string",
        minLength: 3,
        maxLength: 24,
      },
      address: {
        type: "object",
        properties: {
          line1: {
            type: "string",
          },
        },
      },
      cars: {
        type: "object",
        properties: {
          "/.*/": {
            type: "object",
            properties: {
              vin: {
                type: "string",
              },
            },
          },
        },
      },
    },
  };
  const customer: Customer = {
    name: "nish",
    address: {
      line1: "18/19 Lightsview",
    },
    cars: {
      rugby: {
        vin: "xsb797",
      },
    },
  };
  const result = validate(customer, customerSchema);
  expect(result).toMatchSnapshot();
});

test(`Test index type - regex - 2`, () => {
  type Car = {
    vin: string;
  };

  interface Customer {
    name: string;
    address: Address;
    cars: {
      [key: string]: Car;
    };
  }
  interface Address {
    line1: string;
  }
  const customerSchema: Schema<Customer> = {
    type: "object",
    optional: false,
    properties: {
      name: {
        type: "string",
        minLength: 3,
        maxLength: 24,
      },
      address: {
        type: "object",
        properties: {
          line1: {
            type: "string",
          },
        },
      },
      cars: {
        type: "object",
        properties: {
          "/ru.*/": {
            type: "object",
            properties: {
              vin: {
                type: "string",
              },
            },
          },
        },
      },
    },
  };
  const customer: Customer = {
    name: "nish",
    address: {
      line1: "18/19 Lightsview",
    },
    cars: {
      buggy: {
        vin: "", // invalid, but not defined in schema
      },
      ruby: {
        vin: "", // invalid, defined in schema in regex /ru.*/
      },
      rugby: {
        vin: "xsb797",
      },
    },
  };
  const result = validate(customer, customerSchema);
  expect(result).toMatchSnapshot();
});

test(`Test index type - regex - index type is undefined`, () => {
  type Car = {
    vin: string;
  };

  interface Customer {
    name: string;
    cars?: {
      [key: string]: Car;
    };
  }

  const customerSchema: Schema<Customer> = {
    type: "object",
    optional: false,
    properties: {
      name: {
        type: "string",
        minLength: 3,
        maxLength: 24,
      },
      cars: {
        type: "object",
        properties: {
          "/ru.*/": {
            type: "object",
            properties: {
              vin: {
                type: "string",
              },
            },
          },
        },
      },
    },
  };
  const customer: Customer = {
    name: "John",
  };
  const result = validate(customer, customerSchema);
  expect(result).toMatchSnapshot();
});

test(`Test optional flag - array - valid`, () => {
  type Colors = string[];
  type Person = {
    colors?: Colors;
    colorBlind: boolean;
  };
  const personSchema: Schema<Person> = {
    type: "object",
    properties: {
      colorBlind: {
        type: "boolean",
      },
      colors: {
        type: "array",
        optional: (args) => {
          return args.root.colorBlind;
        },
        items: {
          type: "string",
        },
      },
    },
  };
  // valid
  const person: Person = {
    colorBlind: true,
  };
  const result = validate(person, personSchema);
  expect(result.isValid).toBeTruthy();
  // invalid
  const person2: Person = {
    colorBlind: false,
  };
  const result2 = validate(person2, personSchema);
  expect(result2.isValid).toBeFalsy();
});

test(`Test optional flag - boolean`, () => {
  type Person = {
    colorBlind: boolean;
    hasFavColor?: boolean;
  };
  const personSchema: Schema<Person> = {
    type: "object",
    properties: {
      colorBlind: {
        type: "boolean",
      },
      hasFavColor: {
        type: "boolean",
        optional: (args) => args.root.colorBlind, // TODO check why parent is unknown
      },
    },
  };

  // valid
  const person1: Person = {
    colorBlind: true,
  };
  const result1 = validate(person1, personSchema);
  expect(result1.isValid).toBeTruthy();

  // invalid
  const person2: Person = {
    colorBlind: false,
  };
  const result2 = validate(person2, personSchema);
  expect(result2.isValid).toBeFalsy();
});

test(`Test optional flag - string`, () => {
  type Person = {
    colorBlind: boolean;
    favColor?: string;
  };
  const personSchema: Schema<Person> = {
    type: "object",
    properties: {
      colorBlind: {
        type: "boolean",
      },
      favColor: {
        type: "string",
        optional: (args) => args.root.colorBlind,
      },
    },
  };

  // valid
  const person1: Person = {
    colorBlind: true,
  };
  const result1 = validate(person1, personSchema);
  expect(result1.isValid).toBeTruthy();

  // invalid
  const person2: Person = {
    colorBlind: false,
  };
  const result2 = validate(person2, personSchema);
  expect(result2.isValid).toBeFalsy();
});

test(`Test optional flag - string`, () => {
  type Person = {
    colorBlind: boolean;
    favColor?: string;
  };
  const personSchema: Schema<Person> = {
    type: "object",
    properties: {
      colorBlind: {
        type: "boolean",
      },
      favColor: {
        type: "string",
        optional: (args) => args.root.colorBlind,
      },
    },
  };

  // valid
  const person1: Person = {
    colorBlind: true,
  };
  const result1 = validate(person1, personSchema);
  expect(result1.isValid).toBeTruthy();

  // invalid
  const person2: Person = {
    colorBlind: false,
  };
  const result2 = validate(person2, personSchema);
  expect(result2.isValid).toBeFalsy();
});

test(`Test optional flag - date`, () => {
  type Person = {
    isBorn: boolean;
    dob?: Date;
  };
  const personSchema: Schema<Person> = {
    type: "object",
    properties: {
      isBorn: {
        type: "boolean",
      },
      dob: {
        type: "date",
        optional: (args) => args.root.isBorn,
      },
    },
  };

  // valid
  const person1: Person = {
    isBorn: true,
  };
  const result1 = validate(person1, personSchema);
  expect(result1.isValid).toBeTruthy();

  // invalid
  const person2: Person = {
    isBorn: false,
  };
  const result2 = validate(person2, personSchema);
  expect(result2.isValid).toBeFalsy();
});

test(`Test optional flag - number`, () => {
  type Person = {
    isBorn: boolean;
    weight?: number;
  };
  const personSchema: Schema<Person> = {
    type: "object",
    properties: {
      isBorn: {
        type: "boolean",
      },
      weight: {
        type: "number",
        optional: (args) => args.root.isBorn,
      },
    },
  };

  // valid
  const person1: Person = {
    isBorn: true,
  };
  const result1 = validate(person1, personSchema);
  expect(result1.isValid).toBeTruthy();

  // invalid
  const person2: Person = {
    isBorn: false,
  };
  const result2 = validate(person2, personSchema);
  expect(result2.isValid).toBeFalsy();
});

test(`Test optional flag - object`, () => {
  type Person = {
    isBorn: boolean;
    info?: {
      name: string;
      dob: Date;
    };
  };
  const personSchema: Schema<Person> = {
    type: "object",
    properties: {
      isBorn: {
        type: "boolean",
      },
      info: {
        type: "object",
        optional: (args) => args.root.isBorn,
        properties: {
          name: {
            type: "string",
          },
          dob: {
            type: "date",
          },
        },
      },
    },
  };

  // valid
  const person1: Person = {
    isBorn: true,
  };
  const result1 = validate(person1, personSchema);
  expect(result1.isValid).toBeTruthy();

  // invalid
  const person2: Person = {
    isBorn: false,
  };
  const result2 = validate(person2, personSchema);
  expect(result2.isValid).toBeFalsy();
});
