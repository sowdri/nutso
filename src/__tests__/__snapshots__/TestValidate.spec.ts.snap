// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test customer schema 1`] = `
{
  "errorMessage": "",
  "errorPath": [],
  "isValid": true,
  "properties": {
    "address": {
      "errorMessage": "",
      "errorPath": [],
      "isValid": true,
      "properties": {
        "line1": {
          "errorMessage": "",
          "errorPath": [],
          "isValid": true,
        },
      },
    },
    "name": {
      "errorMessage": "",
      "errorPath": [],
      "isValid": true,
    },
  },
}
`;

exports[`Test index type - regex - 2 1`] = `
{
  "errorMessage": "Should not be empty.",
  "errorPath": [
    "cars",
    "ruby",
    "vin",
  ],
  "isValid": false,
  "properties": {
    "address": {
      "errorMessage": "",
      "errorPath": [],
      "isValid": true,
      "properties": {
        "line1": {
          "errorMessage": "",
          "errorPath": [],
          "isValid": true,
        },
      },
    },
    "cars": {
      "errorMessage": "Should not be empty.",
      "errorPath": [
        "ruby",
        "vin",
      ],
      "isValid": false,
      "properties": {
        "ruby": {
          "errorMessage": "Should not be empty.",
          "errorPath": [
            "vin",
          ],
          "isValid": false,
          "properties": {
            "vin": {
              "errorMessage": "Should not be empty.",
              "errorPath": [],
              "isValid": false,
            },
          },
        },
        "rugby": {
          "errorMessage": "",
          "errorPath": [],
          "isValid": true,
          "properties": {
            "vin": {
              "errorMessage": "",
              "errorPath": [],
              "isValid": true,
            },
          },
        },
      },
    },
    "name": {
      "errorMessage": "",
      "errorPath": [],
      "isValid": true,
    },
  },
}
`;

exports[`Test index type - regex - index type is undefined 1`] = `
{
  "errorMessage": "Required field.",
  "errorPath": [
    "cars",
  ],
  "isValid": false,
  "properties": {
    "cars": {
      "errorMessage": "Required field.",
      "errorPath": [],
      "isValid": false,
      "properties": {},
    },
    "name": {
      "errorMessage": "",
      "errorPath": [],
      "isValid": true,
    },
  },
}
`;

exports[`Test index type - regex 1`] = `
{
  "errorMessage": "",
  "errorPath": [],
  "isValid": true,
  "properties": {
    "address": {
      "errorMessage": "",
      "errorPath": [],
      "isValid": true,
      "properties": {
        "line1": {
          "errorMessage": "",
          "errorPath": [],
          "isValid": true,
        },
      },
    },
    "cars": {
      "errorMessage": "",
      "errorPath": [],
      "isValid": true,
      "properties": {
        "rugby": {
          "errorMessage": "",
          "errorPath": [],
          "isValid": true,
          "properties": {
            "vin": {
              "errorMessage": "",
              "errorPath": [],
              "isValid": true,
            },
          },
        },
      },
    },
    "name": {
      "errorMessage": "",
      "errorPath": [],
      "isValid": true,
    },
  },
}
`;

exports[`Test index type - required field - missing 1`] = `
{
  "errorMessage": "Required field.",
  "errorPath": [
    "cars",
    "ruby",
  ],
  "isValid": false,
  "properties": {
    "address": {
      "errorMessage": "",
      "errorPath": [],
      "isValid": true,
      "properties": {
        "line1": {
          "errorMessage": "",
          "errorPath": [],
          "isValid": true,
        },
      },
    },
    "cars": {
      "errorMessage": "Required field.",
      "errorPath": [
        "ruby",
      ],
      "isValid": false,
      "properties": {
        "ruby": {
          "errorMessage": "Required field.",
          "errorPath": [],
          "isValid": false,
          "properties": {},
        },
      },
    },
    "name": {
      "errorMessage": "",
      "errorPath": [],
      "isValid": true,
    },
  },
}
`;

exports[`Test index type 1`] = `
{
  "errorMessage": "",
  "errorPath": [],
  "isValid": true,
  "properties": {
    "address": {
      "errorMessage": "",
      "errorPath": [],
      "isValid": true,
      "properties": {
        "line1": {
          "errorMessage": "",
          "errorPath": [],
          "isValid": true,
        },
      },
    },
    "cars": {
      "errorMessage": "",
      "errorPath": [],
      "isValid": true,
      "properties": {
        "rugby": {
          "errorMessage": "",
          "errorPath": [],
          "isValid": true,
          "properties": {
            "vin": {
              "errorMessage": "",
              "errorPath": [],
              "isValid": true,
            },
          },
        },
      },
    },
    "name": {
      "errorMessage": "",
      "errorPath": [],
      "isValid": true,
    },
  },
}
`;
